dbms/src/Storages/DeltaMerge/Segment.cpp

```diff
@@ -429,81 +433,157 @@ SegmentPtr Segment::restoreSegment( //
-    SegmentMetaInfos segment_infos;
-    while (current_segment_id != 0)
-    {
-        Segment::SegmentMetaInfo segment_info;
-        auto target_id = UniversalPageIdFormat::toFullPageId(
-            UniversalPageIdFormat::toFullPrefix(context.keyspace_id, StorageType::Meta, context.physical_table_id),
-            current_segment_id);
-        auto page = checkpoint_info->temp_ps->read(target_id, nullptr, {}, false);
-        if unlikely (!page.isValid())
+    bool is_cache_ready = false;
+    // If there is a table building cache, then other table may block to read the built cache.
+    // If the remote reader causes much time to retrieve data, then these tasks could block here.
+    // However, when the execlusive holder is canceled due to timeout, the readers could eventually get the lock.
+    FAIL_POINT_PAUSE(FailPoints::pause_when_building_fap_segments);
+    {
+        auto sec = sw.elapsedSecondsFromLastTime();
+        // Lock acquires when:
+        // 1. No writer.
+        // 2. The writer finishes.
+        // 3. The writer is canceled.
+        auto lock = end_to_segment_id_cache->readLock();
+        is_cache_ready = end_to_segment_id_cache->isReady(lock);
+        auto el = sw.elapsedSecondsFromLastTime() - sec;
+        if (el > WAIT_TIME_THRESHOLD)
         {
-            // After #7642, DELTA_MERGE_FIRST_SEGMENT_ID may not exist, however, such checkpoint won't be selected.
-            // If it were to be selected, the FAP task could fallback to regular snapshot.
-            throw Exception(
-                ErrorCodes::LOGICAL_ERROR,
-                "Can't find page id {}, keyspace={} table_id={} current_segment_id={} range={}",
-                target_id,
-                context.keyspace_id,
-                context.physical_table_id,
+            LOG_INFO(
+                Logger::get(),
+                "Wait building segmend id cache for {:.3f}s, current_segment_id={}, region_id={}",
+                el,
                 current_segment_id,
-                target_range.toDebugString());
+                checkpoint_info->region_id);
         }
-        segment_info.segment_id = current_segment_id;
-        ReadBufferFromMemory buf(page.data.begin(), page.data.size());
-        readSegmentMetaInfo(buf, segment_info);
-        if (!is_cache_ready)
+    }
+
+    auto protected_logic = [&](GenericLock && lock) {
+        if (is_cache_ready)
         {
-            end_key_and_segment_ids.emplace_back(segment_info.range.getEnd().toRowKeyValue(), segment_info.segment_id);
+            current_segment_id = end_to_segment_id_cache->getSegmentIdContainingKey(
+                std::get<std::shared_lock<std::shared_mutex>>(lock),
+                target_range.getStart().toRowKeyValue());
         }
-        current_segment_id = segment_info.next_segment_id;
-        if (!(segment_info.range.shrink(target_range).none()))
+        LOG_DEBUG(
+            Logger::get(),
```

Comment:
Why always using `logger::get()` instead of geting the logger at first `auto log = logger::get();`

Reply:
OK, I will fix that

Reply:
fixed

Reply:
You did not push your commit